### ISSUE TITLE
lodash zip signature for arrays of different types

### ISSUE THREAD

**Issue by RedGlow** [DefinitelyTyped/DefinitelyTyped#20682]
- [x] I tried using the `@types/lodash` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bczengel @chrootsu @stepancar @ericanderson @aj-r @Ailrun @e-cloud 

lodash's zip function is able to function on arrays of different types. Taking the documentation example of lodash:

```javascript
_.zip(['a', 'b'], [1, 2], [true, false]);
// => [['a', 1, true], ['b', 2, false]]
```

But the type signature allows only for arrays of the same type:

```typescript
zip<T>(...arrays: Array<List<T> | null | undefined>): T[][];
```

An example like the one before can be forced by using something like `_.zip<any>(...)` or better `_.zip<string | number | bool>(...)`, but this is still a partial solution.

Wouldn't it be better to expand the zip definition also to different array types using Typescript's tuples?

```typescript
zip<T1, T2>(array1: ArrayLike<T1>, array2: ArrayLike<T2>): [T1, T2][];
zip<T1, T2, T3>(array1: ArrayLike<T1>, array2: ArrayLike<T2>, array3: ArrayLike<T3>): [T1, T2, T3][];
// and so on...
```

This solution would have the problem to list a number of possible zips, which is theoretically infinite but can be limited to a predefined amount. I see that already a definition like unionBy follows this approach.

Could there be some compatibility problem with older Typescript versions?

**Comment by aj-r** [comment]
I think it makes sense to add overloads that take tuples.

We don't need to worry about compatibility with Typescript versions older than 2.2.

**Comment by aj-r** [comment]
@andy-ms this can be closed (via #23232)
